### ISSUE TITLE
utest `axpy:zaxpy_inc_0` and `axpy:daxpy_inc_0` failed during compilation v0.3.3 on Intel Atom N2600 CPU

### ISSUE THREAD

**Issue by CC-Hsu** [xianyi/OpenBLAS#1794]
Hi,

I tried to compile OpenBLAS v0.3.3 on Intel Atom N2600 CPU machine, But two of unit test items failed.

Below is the log output with error,
```
make -j 4 -C utest all
make[1]: warning: -jN forced in submake: disabling jobserver mode.
make[1]: Entering directory '/tmp/SBo/OpenBLAS-0.3.3/utest'
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=utest_mainf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=utest_main -DASMFNAME=utest_main_ -DNAME=utest_main_ -DCNAME=utest_main -DCHAR_NAME=\"utest_main_\" -DCHAR_CNAME=\"utest_main\" -DNO_AFFINITY -I..   -c -o utest_main.o utest_main.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_amaxf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_amax -DASMFNAME=test_amax_ -DNAME=test_amax_ -DCNAME=test_amax -DCHAR_NAME=\"test_amax_\" -DCHAR_CNAME=\"test_amax\" -DNO_AFFINITY -I..   -c -o test_amax.o test_amax.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_rotmgf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_rotmg -DASMFNAME=test_rotmg_ -DNAME=test_rotmg_ -DCNAME=test_rotmg -DCHAR_NAME=\"test_rotmg_\" -DCHAR_CNAME=\"test_rotmg\" -DNO_AFFINITY -I..   -c -o test_rotmg.o test_rotmg.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_axpyf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_axpy -DASMFNAME=test_axpy_ -DNAME=test_axpy_ -DCNAME=test_axpy -DCHAR_NAME=\"test_axpy_\" -DCHAR_CNAME=\"test_axpy\" -DNO_AFFINITY -I..   -c -o test_axpy.o test_axpy.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_dotuf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_dotu -DASMFNAME=test_dotu_ -DNAME=test_dotu_ -DCNAME=test_dotu -DCHAR_NAME=\"test_dotu_\" -DCHAR_CNAME=\"test_dotu\" -DNO_AFFINITY -I..   -c -o test_dotu.o test_dotu.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_dsdotf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_dsdot -DASMFNAME=test_dsdot_ -DNAME=test_dsdot_ -DCNAME=test_dsdot -DCHAR_NAME=\"test_dsdot_\" -DCHAR_CNAME=\"test_dsdot\" -DNO_AFFINITY -I..   -c -o test_dsdot.o test_dsdot.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_swapf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_swap -DASMFNAME=test_swap_ -DNAME=test_swap_ -DCNAME=test_swap -DCHAR_NAME=\"test_swap_\" -DCHAR_CNAME=\"test_swap\" -DNO_AFFINITY -I..   -c -o test_swap.o test_swap.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_rotf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_rot -DASMFNAME=test_rot_ -DNAME=test_rot_ -DCNAME=test_rot -DCHAR_NAME=\"test_rot_\" -DCHAR_CNAME=\"test_rot\" -DNO_AFFINITY -I..   -c -o test_rot.o test_rot.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_potrsf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_potrs -DASMFNAME=test_potrs_ -DNAME=test_potrs_ -DCNAME=test_potrs -DCHAR_NAME=\"test_potrs_\" -DCHAR_CNAME=\"test_potrs\" -DNO_AFFINITY -I..   -c -o test_potrs.o test_potrs.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=test_forkf_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME=test_fork -DASMFNAME=test_fork_ -DNAME=test_fork_ -DCNAME=test_fork -DCHAR_NAME=\"test_fork_\" -DCHAR_CNAME=\"test_fork\" -DNO_AFFINITY -I..   -c -o test_fork.o test_fork.c
cc -O2 -DUTEST_CHECK -DSANITY_CHECK -DREFNAME=f_ -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DUSE_SIMPLE_THREADED_LEVEL3 -DUSE_TLS -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -o openblas_utest utest_main.o test_amax.o test_rotmg.o test_axpy.o test_dotu.o test_dsdot.o test_swap.o test_rot.o test_potrs.o test_fork.o ../libopenblas_atomp-r0.3.3.a -lm -lpthread -lgfortran -lm -lpthread -lgfortran -L/usr/lib64/gcc/x86_64-slackware-linux/8.2.0 -L/usr/lib64/gcc/x86_64-slackware-linux/8.2.0/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/usr/lib64/gcc/x86_64-slackware-linux/8.2.0/../../..  -lgfortran -lm -lquadmath -lm -lc  
./openblas_utest
TEST 1/23 amax:samax [OK]
TEST 2/23 drotmg:drotmg_D1_big_D2_big_flag_zero [OK]
TEST 3/23 drotmg:rotmg_D1eqD2_X1eqX2 [OK]
TEST 4/23 drotmg:rotmg_issue1452 [OK]
TEST 5/23 drotmg:rotmg [OK]
TEST 6/23 axpy:caxpy_inc_0 [OK]
TEST 7/23 axpy:saxpy_inc_0 [OK]
TEST 8/23 axpy:zaxpy_inc_0 [FAIL]
  ERR: test_axpy.c:71  expected -3.000e+00, got -5.000e-01 (diff -2.500e+00, tol 1.000e-13)
TEST 9/23 axpy:daxpy_inc_0 [FAIL]
  ERR: test_axpy.c:52  expected 4.000e+00, got 2.250e+00 (diff 1.750e+00, tol 1.000e-13)
TEST 10/23 zdotu:zdotu_offset_1 [OK]
TEST 11/23 zdotu:zdotu_n_1 [OK]
TEST 12/23 dsdot:dsdot_n_1 [OK]
TEST 13/23 swap:cswap_inc_0 [OK]
TEST 14/23 swap:sswap_inc_0 [OK]
TEST 15/23 swap:zswap_inc_0 [OK]
TEST 16/23 swap:dswap_inc_0 [OK]
TEST 17/23 rot:csrot_inc_0 [OK]
TEST 18/23 rot:srot_inc_0 [OK]
TEST 19/23 rot:zdrot_inc_0 [OK]
TEST 20/23 rot:drot_inc_0 [OK]
TEST 21/23 potrf:smoketest_trivial [OK]
TEST 22/23 potrf:bug_695 [OK]
TEST 23/23 fork:safety [OK]
RESULTS: 23 tests (21 ok, 2 failed, 0 skipped) ran in 5173 ms
make[1]: *** [Makefile:33: run_test] Error 2
make[1]: Leaving directory '/tmp/SBo/OpenBLAS-0.3.3/utest'
make: *** [Makefile:125: tests] Error 2
```

Previously v0.2.19 is OK on this machine; I am not sure other versions between v0.2.19 and v0.3.3.

Is this a problem caused by CPU or anything to be fixed in OpenBLAS v0.3.3?

Thanks.

**Comment by martin-frbg** [comment]
This is a sign of missing support for a corner case in the axpy kernel for Atom - most of the utests were added after 0.2.19 so you would not have noticed this before (unless your own code calls DAXPY or ZAXPY with X and Y increments of zero).
The short-term fix would be to comment out the DAXPYKERNEL and ZAXPYKERNEL lines in kernel/x86_64/KERNEL.ATOM, for a proper solution one would probably need to look at how the other assembly implementations handle this.

**Comment by brada4** [comment]
Actually reference BLAS fails the test....

**Comment by brada4** [comment]
Netlib sets 1st result element N times, which serves no purpose in general. Test in particular checks that no change is made getting nonsense input, which sort of breaks bug-for-bug compatibility with netlib.
i'd propose to check for increment==0 in the interface and do nothing, and probably feed that upstream too. I dont know.

**Comment by martin-frbg** [comment]
@brada4 I do not think "bug compatibility with netlib" is our goal here, but indeed this and most other utests compare against stored results from the netlib algorithm and the INC=0 case _has_ been corrected in a few OpenBLAS kernels already. Are you sure reference BLAS produces something other than the precomputed results in x2,y2 when fed x1,y1 ?

**Comment by brada4** [comment]
The pre-calculation is from reference , i am just not very agile with `ar`
Looking at reference implementation there is a shortcut `if incx==0 && incy==0 set n=1` or so.

**Comment by martin-frbg** [comment]
Fixed at the C interface level with #1869
